File: Kalorie/Kalorie/Features/Account/AccountViewModel.swift
//
//  AccountViewModel.swift
//  Kalorie
//
//  Created by Josef Antoni on 07.08.2026.
//

import AuthenticationServices
import Foundation
import GoogleSignIn

final class AccountViewModel: ObservableObject {

    // MARK: - State

    enum State {
        case idle
        case linking
        case deletingAccount
    }

    // MARK: - Properties

    @Published private(set) var state: State = .idle
    @Published var alertItem: AlertItem?
    @Published var showDeleteConfirmation = false
    @Published var isReauthenticateAlertVisible = false

    private let authProvider: any AuthProviderProtocol
    private let signOut: any SignOutUseCaseProtocol
    private let signInWithApple: any SignInWithAppleUseCaseProtocol
    private let signInWithGoogle: any SignInWithGoogleUseCaseProtocol
    private let deleteAccount: any DeleteAccountUseCaseProtocol
    private let reauthenticate: any ReauthenticateUseCaseProtocol
    private let mergeStatusReporting: any MergeStatusReporting
    private var isDataAlreadyWiped = false

    var isAnonymous: Bool { authProvider.isAnonymous }
    var displayName: String? { authProvider.displayName }

    // MARK: - Init

    init(
        authProvider: any AuthProviderProtocol,
        signOut: any SignOutUseCaseProtocol,
        signInWithApple: any SignInWithAppleUseCaseProtocol,
        signInWithGoogle: any SignInWithGoogleUseCaseProtocol,
        deleteAccount: any DeleteAccountUseCaseProtocol,
        reauthenticate: any ReauthenticateUseCaseProtocol,
        mergeStatusReporting: any MergeStatusReporting
    ) {
        self.authProvider = authProvider
        self.signOut = signOut
        self.signInWithApple = signInWithApple
        self.signInWithGoogle = signInWithGoogle
        self.deleteAccount = deleteAccount
        self.reauthenticate = reauthenticate
        self.mergeStatusReporting = mergeStatusReporting
    }

    // MARK: - Functions

    func onSignOutTapped() {
        do {
            try signOut()
        } catch {
            Log.error(error, category: Constants.LogCategory.account)
            alertItem = AlertItem(title: L10n.Account.errorSignOutFailed)
        }
    }

    @MainActor
    func onSignInWithAppleTapped() async {
        state = .linking
        mergeStatusReporting.beginMerge()
        defer {
            mergeStatusReporting.endMerge()
            state = .idle
        }
        do {
            try await signInWithApple()
        } catch where isUserCancellation(error) {
        } catch {
            Log.error(error, category: Constants.LogCategory.account)
            alertItem = AlertItem(title: L10n.Account.errorSignInFailed)
        }
    }

    @MainActor
    func onSignInWithGoogleTapped() async {
        state = .linking
        mergeStatusReporting.beginMerge()
        defer {
            mergeStatusReporting.endMerge()
            state = .idle
        }
        do {
            try await signInWithGoogle()
        } catch LinkOrMergeCredentialError.accountExistsWithAnotherProvider {
            alertItem = AlertItem(title: L10n.Account.errorAccountExistsWithApple)
        } catch where isUserCancellation(error) {
        } catch {
            Log.error(error, category: Constants.LogCategory.account)
            alertItem = AlertItem(title: L10n.Account.errorSignInFailed)
        }
    }

    @MainActor
    func onDeleteAccountConfirmed() async {
        isDataAlreadyWiped = false
        defer { state = .idle }
        await performDelete()
    }

    @MainActor
    func onReauthenticateConfirmed() async {
        state = .linking
        defer { state = .idle }
        do {
            try await reauthenticate()
            await performDelete()
        } catch where isUserCancellation(error) {
        } catch {
            Log.error(error, category: Constants.LogCategory.account)
            alertItem = AlertItem(title: L10n.Account.errorSignInFailed)
        }
    }

    // MARK: - Private

    private func isUserCancellation(_ error: Error) -> Bool {
        if
            let authError = error as? ASAuthorizationError,
            authError.code == .canceled
        {
            return true
        }
        let nsError = error as NSError
        return nsError.domain == GIDSignInError.errorDomain && nsError.code == GIDSignInError.canceled.rawValue
    }

    @MainActor
    private func performDelete() async {
        state = .deletingAccount
        do {
            try await deleteAccount(skipDataWipe: isDataAlreadyWiped)
        } catch DeleteAccountError.requiresRecentLogin(let dataAlreadyDeleted) {
            isDataAlreadyWiped = dataAlreadyDeleted
            isReauthenticateAlertVisible = true
        } catch {
            Log.error(error, category: Constants.LogCategory.account)
            alertItem = AlertItem(title: L10n.Account.errorDeleteFailed)
        }
    }
}

File: Android/app/src/main/java/antoni/kalorie/features/account/AccountViewModel.kt
package antoni.kalorie.features.account

import androidx.credentials.exceptions.GetCredentialCancellationException
import androidx.lifecycle.ViewModel
import antoni.kalorie.R
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.auth.MergeStatusReporting
import antoni.kalorie.core.usecases.DeleteAccountError
import antoni.kalorie.core.usecases.DeleteAccountUseCaseProtocol
import antoni.kalorie.core.usecases.FetchMaintainerClaimUseCaseProtocol
import antoni.kalorie.core.usecases.LinkOrMergeCredentialError
import antoni.kalorie.core.usecases.ReauthenticateUseCaseProtocol
import antoni.kalorie.core.usecases.SignInWithGoogleUseCaseProtocol
import antoni.kalorie.core.usecases.SignOutUseCaseProtocol
import antoni.kalorie.core.utils.AlertItem
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.Log
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

class AccountViewModel(
    private val authProvider: AuthProviderProtocol,
    private val signOut: SignOutUseCaseProtocol,
    private val signInWithGoogle: SignInWithGoogleUseCaseProtocol,
    private val deleteAccount: DeleteAccountUseCaseProtocol,
    private val reauthenticate: ReauthenticateUseCaseProtocol,
    private val fetchMaintainerClaim: FetchMaintainerClaimUseCaseProtocol,
    private val mergeStatusReporting: MergeStatusReporting,
) : ViewModel() {

    // MARK: - State

    enum class State {
        IDLE,
        LINKING,
        DELETING_ACCOUNT,
    }

    // MARK: - Properties

    private val _state = MutableStateFlow(State.IDLE)
    val state: StateFlow<State> = _state
    val alertItem = MutableStateFlow<AlertItem?>(null)
    val showDeleteConfirmation = MutableStateFlow(false)
    val isReauthenticateAlertVisible = MutableStateFlow(false)
    private val _isMaintainer = MutableStateFlow(false)
    val isMaintainer: StateFlow<Boolean> = _isMaintainer
    private var isDataAlreadyWiped = false

    private val _isAnonymous = MutableStateFlow(authProvider.isAnonymous)
    val isAnonymous: StateFlow<Boolean> = _isAnonymous
    private val _displayName = MutableStateFlow(authProvider.displayName)
    val displayName: StateFlow<String?> = _displayName

    // MARK: - Functions

    suspend fun onAppear() {
        try {
            _isMaintainer.value = fetchMaintainerClaim()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.warning(error, Constants.LogCategory.ACCOUNT)
        }
    }

    suspend fun onSignOutTapped() {
        try {
            signOut()
            _isMaintainer.value = false
            refreshAuthState()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.ACCOUNT)
            alertItem.value = AlertItem(titleRes = R.string.account_error_signOutFailed)
        }
    }

    suspend fun onSignInWithGoogleTapped() {
        _state.value = State.LINKING
        mergeStatusReporting.beginMerge()
        try {
            signInWithGoogle()
        } catch (error: CancellationException) {
            throw error
        } catch (_: LinkOrMergeCredentialError.AccountExistsWithAnotherProvider) {
            alertItem.value = AlertItem(titleRes = R.string.account_error_accountExistsWithApple)
        } catch (error: Exception) {
            if (!isUserCancellation(error)) {
                Log.error(error, Constants.LogCategory.ACCOUNT)
                alertItem.value = AlertItem(titleRes = R.string.account_error_signInFailed)
            }
        } finally {
            mergeStatusReporting.endMerge()
            refreshAuthState()
            _state.value = State.IDLE
        }
    }

    suspend fun onDeleteAccountConfirmed() {
        isDataAlreadyWiped = false
        try {
            performDelete()
        } finally {
            refreshAuthState()
            _state.value = State.IDLE
        }
    }

    suspend fun onReauthenticateConfirmed() {
        _state.value = State.LINKING
        try {
            reauthenticate()
            performDelete()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            if (!isUserCancellation(error)) {
                Log.error(error, Constants.LogCategory.ACCOUNT)
                alertItem.value = AlertItem(titleRes = R.string.account_error_signInFailed)
            }
        } finally {
            refreshAuthState()
            _state.value = State.IDLE
        }
    }

    // MARK: - Private

    private fun refreshAuthState() {
        _isAnonymous.value = authProvider.isAnonymous
        _displayName.value = authProvider.displayName
    }

    private fun isUserCancellation(error: Throwable): Boolean = error is GetCredentialCancellationException

    private suspend fun performDelete() {
        _state.value = State.DELETING_ACCOUNT
        try {
            deleteAccount(isDataAlreadyWiped)
        } catch (error: CancellationException) {
            throw error
        } catch (error: DeleteAccountError.RequiresRecentLogin) {
            isDataAlreadyWiped = error.dataAlreadyDeleted
            isReauthenticateAlertVisible.value = true
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.ACCOUNT)
            alertItem.value = AlertItem(titleRes = R.string.account_error_deleteFailed)
        }
    }
}
